File: Data Structures & Algorithms/top-k-elements-in-list/submission-1.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} k
     * @return {number[]}
     */
    topKFrequent(nums, k) {
        let max = Math.max(...nums);

        let freq = new Map();

        for(const n of nums){
            freq.set(n, (freq.get(n) || 0) + 1);
        }

        let buckets = Array.from({length: Math.max(...Array.from(freq.values())) + 1}, () => []);

        for(const [n, f] of freq.entries()){
            buckets[f].push(n);
        }


        let out = [];
        for(let i = buckets.length - 1; i >= 0 && out.length < k; i--){
            out = out.concat(buckets[i]);
        }

        return out;
    }
}

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-1.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        cache = {}

        def findPath(i):
            if i >= len(nums):
                return 0

            if i in cache:
                return cache[i]

            m = 1

            for j in range(i + 1, len(nums)):
                if nums[i] < nums[j]:
                    m = max(m, 1 + findPath(j))
            
            cache[i] = m
            return m
            


        m = 0
        for i in range(len(nums)):
            m = max(findPath(i), m)

        return m